api4/user.go

```diff
@@ -1129,6 +1140,93 @@ func updateUser(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(ruser.ToJson()))
 }
 
+func patchUserCheckLoginProviderAttributes(c *Context, ouser *model.User, patch *model.UserPatch) {
+	const (
+		firstNameField = "first_name"
+		lastNameField  = "last_name"
+		nicknameField  = "nickname"
+		positionField  = "position"
+		emailField     = "email"
+		usernameField  = "username"
+	)
+	// If LDAP or SAML is used, and any of the following attributes have been specified
+	// in the settings, then their corresponding fields may not be changed
+	LS := &c.App.Config().LdapSettings
+	SS := &c.App.Config().SamlSettings
+	for _, provider := range []struct {
+		AuthService        string
+		FirstNameAttribute *string
+		LastNameAttribute  *string
+		NicknameAttribute  *string
+		PositionAttribute  *string
+		EmailAttribute     *string
+	}{
+		{model.USER_AUTH_SERVICE_LDAP, LS.FirstNameAttribute, LS.LastNameAttribute, LS.NicknameAttribute,
+			LS.PositionAttribute, LS.EmailAttribute},
+		{model.USER_AUTH_SERVICE_SAML, SS.FirstNameAttribute, SS.LastNameAttribute, SS.NicknameAttribute,
+			SS.PositionAttribute, SS.EmailAttribute},
+	} {
+		if ouser.AuthService != provider.AuthService {
+			continue
+		}
+		for _, attr := range []struct {
+			name       string
+			patchValue *string
+			userValue  *string
+			selector   *string
+		}{
+			{firstNameField, patch.FirstName, &ouser.FirstName, provider.FirstNameAttribute},
+			{lastNameField, patch.LastName, &ouser.LastName, provider.LastNameAttribute},
+			{nicknameField, patch.Nickname, &ouser.Nickname, provider.NicknameAttribute},
+			{positionField, patch.Position, &ouser.Position, provider.PositionAttribute},
+			{emailField, patch.Email, &ouser.Email, provider.EmailAttribute},
+		} {
+			if attr.patchValue != nil && *attr.patchValue != *attr.userValue && *attr.selector != "" {
+				c.Err = model.NewAppError(
+					"patchUser", "api.user.patch_user.login_provider_attribute_set.app_error",
+					map[string]interface{}{"Field": attr.name}, "", http.StatusConflict)
+				return
+			}
+		}
+	}
+	// If any login provider is used, then the username may not be changed
+	if ouser.AuthService != "" && patch.Username != nil && *patch.Username != ouser.Username {
+		c.Err = model.NewAppError(
+			"patchUser", "api.user.patch_user.login_provider_attribute_set.app_error",
+			map[string]interface{}{"Field": usernameField}, "",
+			http.StatusConflict)
+		return
+	}
+	// If an OAUTH provider is used then neither the first name nor the last
+	// name may be changed
+	for _, oauthService := range []string{
+		model.USER_AUTH_SERVICE_GITLAB,
+		"google",
+		"office365",
+		"openid",
+	} {
+		if ouser.AuthService != oauthService {
+			continue
+		}
+		for _, attr := range []struct {
+			name       string
+			patchValue *string
+			userValue  *string
+		}{
+			{firstNameField, patch.FirstName, &ouser.FirstName},
+			{lastNameField, patch.FirstName, &ouser.LastName},
```

Comment:
```suggestion
			{lastNameField, patch.LastName, &ouser.LastName},
```
Do you think we should add a unit test to ensure this is covered?

Reply:
Unit tests for LDAP users are in the enterprise repo. Tests for SAML and OAUTH users will be added once we figure out how to run Keycloak during our tests.